homeassistant/components/rest/__init__.py

```diff
@@ -1,4 +1,170 @@
+
+_LOGGER = logging.getLogger(__name__)
+
 PLATFORMS = ["binary_sensor", "notify", "sensor", "switch"]
+COORDINATOR_AWARE_PLATFORMS = [SENSOR_DOMAIN, BINARY_SENSOR_DOMAIN]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the rest platforms."""
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
+    _async_setup_shared_data(hass)
+
+    async def reload_service_handler(service):
+        """Remove all user-defined groups and load new ones from config."""
+        conf = await component.async_prepare_reload()
+        if conf is None:
+            return
+        await async_reload_integration_platforms(hass, DOMAIN, PLATFORMS)
+        _async_setup_shared_data(hass)
+        await _async_process_config(hass, conf)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_RELOAD, reload_service_handler, schema=vol.Schema({})
+    )
+
+    return await _async_process_config(hass, config)
+
+
+@callback
+def _async_setup_shared_data(hass: HomeAssistant):
+    """Create shared data for platform config and rest coordinators."""
+    hass.data[DOMAIN] = {platform: {} for platform in COORDINATOR_AWARE_PLATFORMS}
+
+
+async def _async_process_config(hass, config) -> bool:
+    """Process rest configuration."""
+    if DOMAIN not in config:
+        return True
+
+    load_tasks = []
+    for rest_idx, conf in enumerate(config[DOMAIN]):
+        scan_interval = conf.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+        resource_template = conf.get(CONF_RESOURCE_TEMPLATE)
+        rest = create_rest_data_from_config(hass, conf)
+        coordinator = _wrap_rest_in_coordinator(
+            hass, rest, resource_template, scan_interval
+        )
+        await coordinator.async_refresh()
+        hass.data[DOMAIN][rest_idx] = {REST: rest, COORDINATOR: coordinator}
+
+        for platform_domain in COORDINATOR_AWARE_PLATFORMS:
+            if platform_domain not in conf:
+                continue
+
+            for platform_idx, platform_conf in enumerate(conf[platform_domain]):
+                hass.data[DOMAIN][platform_domain][platform_idx] = platform_conf
+
+                load = discovery.async_load_platform(
+                    hass,
+                    platform_domain,
+                    DOMAIN,
+                    {REST_IDX: rest_idx, PLATFORM_IDX: platform_idx},
+                    config,
+                )
+                load_tasks.append(load)
+
+    if load_tasks:
+        await asyncio.gather(*load_tasks)
+
+    return True
+
+
+async def async_get_config_and_coordinator(hass, platform_domain, discovery_info):
+    """Get the config and coordinator for the platform from discovery."""
+    shared_data = hass.data[DOMAIN][discovery_info[REST_IDX]]
+    conf = hass.data[DOMAIN][platform_domain][discovery_info[PLATFORM_IDX]]
+    coordinator = shared_data[COORDINATOR]
+    rest = shared_data[REST]
+    if rest.data is None:
+        await coordinator.async_request_refresh()
```

Comment:
Is this needed? It looks like `_async_process_config` already makes sure we have data ?

Reply:
Yes, if the platform fails to setup the first time it needs to try again.

Reply:
Weird. This code `prepare_setup_platform` is called by EntityComponent and so the integration should be set up: 

https://github.com/home-assistant/core/blob/dev/homeassistant/setup.py#L269-L309

Reply:
ok, I need to go to sleep. I thought you responded to another comment of mine.

Reply:
```
   if rest.data is None:
        raise PlatformNotReady
```
Its to handle this case more specifically 